src/main/webapp/WEB-INF/view/chat.jsp

```diff
@@ -34,19 +38,77 @@ List<Message> messages = (List<Message>) request.getAttribute("messages");
   <script src="https://js.pusher.com/4.1/pusher.min.js"></script>
 
   <script type="text/javascript">
+    // Pusher code
     var pusher = new Pusher('${PusherProvider.PUSHER_KEY}', {
       cluster: 'us2',
       encrypted: true
     });
 
     var channel = pusher.subscribe('${PusherProvider.CHAT_CHANNEL}');
 
+    // When notification received
     channel.bind('${PusherProvider.MESSAGE_SENT}', function (data) {
-      var list = document.getElementById('message-list');
-      list.innerHTML += '<li>' + data.view + '</li>';
+      messages[data.id] = {};
+      messages[data.id].view = data.view;
+      messages[data.id].children = [];
+      messages[data.id].depth = data.parentId == "" ? 0 : messages[data.parentId].depth+1;
+      var list = document.getElementById('message-list'+data.parentId);
+      list.innerHTML += formatMessageViewGroup(data.id, "");
       scrollChat();
     });
 
+    var messages = {};
+    var orderedRootMessages = [];
+
+    function formatMessageViewGroup(messageId, childrenHtml) {
+      var message = messages[messageId];
+      return "<li style=\"padding-left:" + message.depth*12 + "px;\">" + message.view + "</li>"
```

Comment:
The style stuff should probably be in the .css files to keep it all in the same place. There's a chat.css file that this could go in or it can stay here for now